File: python/151_ReverseWords.py
"""
Given an input string, reverse the string word by word.

For example,
Given s = "the sky is blue",
return "blue is sky the".
"""


class Solution(object):
    def reverseWords(self, s):
        """
        :type s: a list of 1 length strings (List[str])
        :rtype: nothing
        """
        def reverse(start, end, s):
            while start < end:
                s[start], s[end] = s[end], s[start]
                start += 1
                end -= 1
        reverse(0, len(s)-1, s)
        p = 0
        for q, c in enumerate(s):
            if c == " ":
                reverse(p, q, s[p:q])
                p = q + 1
        return s





s = Solution()
print s.reverseWords("the sky is blue")

File: golang/151_ReverseWords.go
/*
Given an input string, reverse the string word by word.

For example,
Given s = "the sky is blue",
return "blue is sky the".
*/

package main
import (
    "fmt"
)
func reverse(s string, start int, end int)(string) {
    r := []rune(s)
    for i, j := start, end; i < j; i, j = i+1, j-1 {
        r[i], r[j] = r[j], r[i]
    }
    return string(r)
}

func reverseWords(s string)(string) {
    reversed := reverse(s, 0, len(s) - 1)
    p := 0
    for q, char := range reversed {
        if char == " " {
            reverse(s[p:q], p, q)
            p = q + 1
        }
    }
}
